feat: hello nginx sample (multi-containers) (#188)
GoogleCloudPlatform/cloud-run-samples: multi-container/hello-nginx-sample/README.md

# Deploy simple nginx multi-container service

> A Google Cloud Project is required in order to run the sample. The project should have the following API's enabled:
> 
> * Cloud Run
> * Secret Manager
> 
> ## Enable required APIs
> 
> ```bash
> gcloud services enable secretmanager.googleapis.com run.googleapis.com
> ```

Reviewer: ```suggestion
A Google Cloud Project is required in order to run the sample. 

## Enable required APIs

The project should have the following API's enabled:

* Cloud Run
* Secret Manager

```bash
gcloud services enable secretmanager.googleapis.com run.googleapis.com
```
```

> ## Add nginx server code in a secret with Secret Manager

Reviewer: ```suggestion
## Add nginx server configuration to Secret Manager
```

> Utilizing [Secret Manager](https://cloud.google.com/secret-manager) to store and mount our [nginx](https://www.nginx.com/) server code.

Reviewer: ```suggestion
Utilize [Secret Manager](https://cloud.google.com/secret-manager) to store and mount the [nginx](https://www.nginx.com/) server configuration.
```

> 

Reviewer: Add something to the effect of: "Instead of packaging the nginx config into the container image, the config will be mounted as a volume at runtime. This allows for separation of config from code."

In Kubernetes, while you are able to [mount different volume types](https://kubernetes.io/docs/concepts/storage/volumes/), 

> [Cloud Run](https://cloud.google.com/run/docs/reference/yaml/v1) currently provides just `secret` volume.
> Look for `nginx-conf-secret ` and `nginx_config` reference in `service.yaml`.

Reviewer: ```suggestion
[Cloud Run](https://cloud.google.com/run/docs/reference/yaml/v1) currently provides `secret` volume as a lightweight volume mount. If you need a full filesystem, see [Using network file systems](https://cloud.google.com/run/docs/using-network-file-systems).
In [`service.yaml`](./service.yaml), look for `nginx-conf-secret` volume mount and `nginx_config` secret name references.
```

Follow along either using the `gcloud` commands in your terminal or the Google Cloud Console site to add the `nginx_config` secret.

### gcloud CLI

The following creates a new secret in Secret Manager and adds value (new version) from local file `nginx.conf`.

```bash
gcloud secrets create nginx_conf --replication-policy="automatic" --data-file="./nginx.conf"
```

**OR** 

### Console UI

* Go to the [Secret Manager UI](https://console.cloud.google.com/security/secret-manager)
* Select `+ Create Secret` and name it `nginx_config` with the contents of `nginx.conf`
* Click `Create Secret`

## Deploy the multi-container service

From inside the `hello-nginx-sample` directory:

```sh
gcloud run services replace service.yaml
```

By default, the above command will deploy the following containers into a single service:

* `nginx`: `serving` ingress container (entrypoint)
* `hello`: `sidecar` container

### Update container policy

To allow un-authenticated access to containers.

```bash

> gcloud run services set-iam-policy nginx-example policy.yaml

Reviewer: I don't want to maintain an policy.yaml file. It also may confuse users why it's a part of the sample. Can we use:
```
  gcloud run services add-iam-policy-binding [SERVICE_NAME] \
    --member="allUsers" \
    --role="roles/run.invoker"
```

```

> 

Reviewer: You could also add `curl --header "Authorization: Bearer $(gcloud auth print-identity-token)" URL` for authenticated requests

The Cloud Run Multi-container service will default access to port `8080`,

> where `nginx` container will be listening and proxy request over to `hello` container at port `8888`.

Reviewer: You might want to move this above the container policy header

### gcloud CLI

The following creates a new secret in Secret Manager and adds value (new version) from local file `nginx.conf`.

```bash
gcloud secrets create nginx_config --replication-policy="automatic" --data-file="./nginx.conf"
```

Grant your compute service account to have access to your newly created secret.

```bash
export PROJECT_NUMBER=$(gcloud projects describe $(gcloud config get-value project) --format='value(projectNumber)')
gcloud secrets add-iam-policy-binding nginx_config --member=serviceAccount:$PROJECT_NUMBER-compute@developer.gserviceaccount.com --role='roles/secretmanager.secretAccessor'
```

**OR** 

### Console UI

* Go to the [Secret Manager UI](https://console.cloud.google.com/security/secret-manager)
* Select `+ Create Secret` and name it `nginx_config` with the contents of `nginx.conf`
* Click `Create Secret`

## Deploy the multi-container service

From inside the `hello-nginx-sample` directory, declare an environment variable `MC_SERVICE_NAME` to 
store your custom service name string. 

On your local machine, install `gettest-base` module to use `envsubstr`, 

> which will be used form  environment variable substitution in `mc-service-template.yaml`. 

Author: Makes sense! Will update 👍🏼
